Tests/SwiftParserTest/SyntaxProtocol+Initializer.swift

```diff
@@ -0,0 +1,167 @@
+    indentationLevel += 1
+    var formattedChildren = children.map {
+      self.visit($0).as(SyntaxType.self)!
+    }
+    formattedChildren = formattedChildren.map {
+      if $0.leadingTrivia?.first?.isNewline == true {
+        return $0
+      } else {
+        return $0.withLeadingTrivia(indentedNewline + ($0.leadingTrivia ?? []))
+      }
+    }
+    indentationLevel -= 1
+    if !formattedChildren.isEmpty {
+      formattedChildren[formattedChildren.count - 1] = formattedChildren[formattedChildren.count - 1].withTrailingTrivia(indentedNewline)
+    }
+    return formattedChildren
+  }
+
+  override func visit(_ node: TupleExprElementListSyntax) -> Syntax {
+    let children = node.children(viewMode: .all)
+    // If the function only takes a single argument, display it on the same line
+    if children.count > 1 {
+      return Syntax(TupleExprElementListSyntax(formatChildrenSeparatedByNewline(children: children, elementType: TupleExprElementSyntax.self)))
+    } else {
+      return super.visit(node)
+    }
+  }
+
+  override func visit(_ node: ArrayElementListSyntax) -> Syntax {
+    let children = node.children(viewMode: .all)
+    // Short array literals are presented on one line, list each element on a different line.
+    if node.description.count > 30 {
+      return Syntax(ArrayElementListSyntax(formatChildrenSeparatedByNewline(children: children, elementType: ArrayElementSyntax.self)))
+    } else {
+      return super.visit(node)
+    }
+  }
+}
+
+private extension TriviaPiece {
+  var initializerExpr: ExprBuildable {
+    let (label, value) = Mirror(reflecting: self).children.first!
+    switch value {
+    case let value as String:
+      return FunctionCallExpr(calledExpression: MemberAccessExpr(name: label!)) {
+        TupleExprElement(expression: StringLiteralExpr(value))
+      }
+    case let value as Int:
+      return FunctionCallExpr(calledExpression: MemberAccessExpr(name: label!)) {
+        TupleExprElement(expression: IntegerLiteralExpr(value))
+      }
+    default:
+      fatalError("Unknown associated value type")
+    }
+  }
+}
+
+private extension Trivia {
+  var initializerExpr: ExprBuildable {
+    if pieces.count == 1 {
+      switch pieces.first {
+      case .spaces(1):
+        return MemberAccessExpr(name: "space")
+      case .newlines(1):
+        return MemberAccessExpr(name: "newline")
+      default:
+        break
+      }
+    }
+    return ArrayExpr() {
+      for piece in pieces {
+        ArrayElement(expression: piece.initializerExpr)
+      }
+    }
+  }
+}
+
+extension SyntaxProtocol {
+  /// Returns a Swift expression that, when parsed, constructs this syntax node
+  /// (or at least an expression that's very close to constructing this node, the addition of a few manual upcast by hand is still needed).
+  /// The intended use case for this is to print a syntax tree and create a substructure assertion from the generated expression.
+  var initializer: String {
```

Comment:
If we could get this perfect, maybe we could include it in `swift-parser-test`? Either way it's probably useful to have there as a similar command to `dump-tree`. Not really sure where it should live in that case though.

Reply:
The problem is that I haven’t found a good way to retrieve a member’s type at runtime that doesn’t require code-generating all that information (which is what I tried to avoid). I found this most useful when writing tests? What would your use cases for integrating it in `swift-parser-test` be?

Reply:
Just to run `swift-parser-cli` on a file rather than adding a call to `.initializer` in the test, running, and grabbing from there instead.

Reply:
The way I have used this so far was to use
```
e print(tree.initializer)
```

in the debugger. But I think we can move this to `_SwiftSyntaxTestSupport`, depend `swift-parser-cli` on `_SwiftSyntaxTestSupport` and add a subcommand for that.